### PR TITLE
Add support for curator for Elasticsearch index management

### PR DESCRIPTION
**Pull request by richm**
Curator allows the admin to remove old data from Elasticsearch
on a per-project basis.  It runs as a separate pod within the
logging project.  See the README.md for more details about how
it is configured and run.

### CODE REVIEW

**Comment by ewolinetz** [comment]
LGTM, but I helped work on it.  Second opinion @sosiouxme @jcantrill  ?

File: README.md [hunk 1]
```diff
@@ -39,7 +39,54 @@ proxy is required that runs in front of Kibana.
 The deployer enables the user to generate all of the necessary
 key/certs/secrets and deploy all of the components in concert.
 
-## Defining local builds
+#### Curator
```
Comment (jcantrill) [inline README.md]: It's not obvious to me what happens in the default scenario.  The case where there is an empty hash for INDEX_MGMT.  The way it reads is that I will need to add an entry for EVERY project which is not what we want.  I would add  a section here that makes it more obvious that doing nothing still gets you deleted indexes.  Also, are project names evaluated based on a wild card since the actual indexes are a combination of a date and project?

**Comment by ewolinetz** [comment]
Should we have this avoid our .searchguard index?  It already avoids the .kibana indices

**Comment by jcantrill** [comment]
probably, though i dont recall what is stashed in that index.

On Tue, Feb 9, 2016 at 2:19 PM, Eric Wolinetz <notifications@github.com>
wrote:

> Should we have this avoid our .searchguard index? It already avoids the
> .kibana indices
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin-aggregated-logging/pull/57#issuecomment-182016018>
> .
>



-- 
--
Jeff Cantrill
Senior Software Engineer, Red Hat Engineering
Red Hat, Inc.
*Office*: 703-748-4420 | 866-546-8970 ext. 8162420
jcantril@redhat.com
http://www.redhat.com

**Comment by ewolinetz** [comment]
@jcantrill I think that is where the ACL info is stored

File: README.md (hunk 1 above)
Comment (ewolinetz) [inline README.md]: correct, we evaluate project names with ".*" appended to account for the date format at the end, and the inability for project names to contain a period (when created through OpenShift)

**Comment by richm** [comment]
The curator configuration needs to change.

curator delete indices --time default_time --unit default_unit --timestring x.y.z --all-indices --except '^.kibana*' --except '^.searchguard*' --except '^project1.*' --except '^project2.*' ...

curator delete indices --time time1 --unit unit1 --timestring x.y.z --prefix 'project1.' --prefix 'project2.'

curator delete indices --time time2 --unit unit2 --timestring x.y.z --prefix 'project3.' --prefix 'project3.'

The default should work on all indices except kibana, searchguard, and any projects for which we have specific trimming times/units.  Each group of projects that share a common time/unit will have their own curator run listing each project in the prefix.

File: elasticsearch/install.sh [hunk 2]
```diff
@@ -15,7 +16,7 @@ export CLUSTER_NAME=placeholder
 mkdir -p ${HOME}
 ln -s /usr/share/elasticsearch /usr/share/java/elasticsearch
 /usr/share/elasticsearch/bin/plugin -i com.floragunn/search-guard/0.5
-/usr/share/elasticsearch/bin/plugin -i io.fabric8.elasticsearch/openshift-elasticsearch-plugin/0.7
+/usr/share/elasticsearch/bin/plugin -i io.fabric8.elasticsearch/openshift-elasticsearch-plugin/0.8
```
Comment (ewolinetz) [inline elasticsearch/install.sh]: since we just released version 0.9 of this plugin could you update to use that version instead?

Comment (richm) [inline elasticsearch/install.sh]: @ewolinetz That should be a separate commit, and I should rebase my commit on that commit

Comment (ewolinetz) [inline elasticsearch/install.sh]: @richm in that case I'd say we can do that as a follow up PR after this is merged in. I don't see a need to wait on another commit to rebase on top of.

**Comment by ewolinetz** [comment]
@richm Do we still need to update the curator configuration as you indicated in your above comment?

**Comment by richm** [comment]
@ewolinetz I think so - I'm testing it now

**Comment by ewolinetz** [comment]
other than my question LGTM

File: curator/install.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+yum install -y --setopt=tsflags=nodocs \
+  python-pip \
+  PyYAML
+pip install elasticsearch-curator python-crontab
+yum clean all
+
+# HACK HACK HACK - remove when fixed upstream
+yum install -y --setopt=tsflags=nodocs \
+    git
+git clone -b issue-344 https://github.com/richm/elasticsearch-py.git
```
Comment (sosiouxme) [inline curator/install.sh]: good gorm that's wretched... i hope we have a more sanctioned way to apply this patch for enterprise?

File: README.md [hunk 4]
```diff
@@ -39,7 +39,71 @@ proxy is required that runs in front of Kibana.
...
+       days: 1
+
+    myapp-qe:
+      delete:
+        weeks: 1
+     ...
+
+Every day, curator will run, and will delete indices in the myapp-dev
+project older than 1 day, and indices in the myapp-qe project older than 1
+week.
```
Comment (sosiouxme) [inline README.md]: Can we call out how to configure curating ops logs? I assume you just use project name `.operations` in the yaml but I think we need to say that.

File: deployment/README.md [hunk 5]
```diff
@@ -31,6 +31,11 @@ The deployer enables the system administrator to generate all of the
 necessary key/certs/secrets and deploy all of the logging components
 in concert.
 
+## Curator
+
+Curator allows the admin to remove old data from Elasticsearch on a per-project
```
Comment (sosiouxme) [inline deployment/README.md]: could we say "and is configurable on a per-project basis"?

File: elasticsearch/install.sh [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 set -ex
 
+yum -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
```
Comment (sosiouxme) [inline elasticsearch/install.sh]: what did we need to add epel for?

File: curator/install.sh (hunk 3 above)
Comment (richm) [inline curator/install.sh]: Yes.  For enterprise we will have to go through the usual approved patch process.  It's going to be a pain anyway, since we're using epel and/or pip packages.

File: deployment/README.md (hunk 5 above)
Comment (richm) [inline deployment/README.md]: done

File: elasticsearch/install.sh (hunk 6 above)
Comment (richm) [inline elasticsearch/install.sh]: I think that is a holdover for when we had curator as part of the elasticsearch image.  I will remove that.

File: deployment/templates/curator.yaml [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+      triggers:
+      - type: ConfigChange
+      - type: ImageChange
+        imageChangeParams:
+          automatic: true
+          containerNames:
+          - curator
+          from:
+            kind: ImageStreamTag
+            name: logging-curator:latest
```
Comment (sosiouxme) [inline deployment/templates/curator.yaml]: This is only going to exist if you've created the ImageStream, which looks to be missing - should be added to deployment/templates/support.yaml I think for deploying from official registry.

Comment (richm) [inline deployment/templates/curator.yaml]: Ok - check the latest commit - does that look correct?

File: deployment/README.md (hunk 5 above)
Comment (sosiouxme) [inline deployment/README.md]: :+1:

File: deployment/templates/curator.yaml (hunk 7 above)
Comment (sosiouxme) [inline deployment/templates/curator.yaml]: :+1:

**Comment by sosiouxme** [comment]
minor comments, LGTM

**Comment by ewolinetz** [comment]
comments look to be addressed, merging
